lib/tasks/install-blueprint.js

```diff
@@ -38,22 +45,99 @@ class InstallBlueprintTask extends Task {
-        let execArgs = ['git', 'clone', blueprintOption, pathName].join(' ');
-        return exec(execArgs).then(() => exec('npm install', { cwd: pathName }).then(() => {
-          let blueprint = Blueprint.load(pathName);
-          return blueprint.install(installOptions);
-        }));
-      });
-    } else {
-      let blueprintName = blueprintOption || 'app';
-      let blueprint = Blueprint.lookup(blueprintName, {
-        paths: this.project.blueprintLookupPaths(),
-      });
-      return blueprint.install(installOptions);
+    return lookupBlueprint(blueprintName, blueprintPaths)
+      .then(blueprint => blueprint.install(installOptions));
+  }
+}
+
+// Given a blueprint name and set of paths to search, looks up the specified
+// blueprint. If it detects a git repository URL, it will clone the repository
+// and use that as the blueprint. If it detects an npm package name, it will
+// install the package and use that as the blueprint.
+function lookupBlueprint(blueprintName, paths) {
+  // Fast path: if no blueprint is defined, we can fall back to the default app
+  // blueprint which we know is available locally.
+  if (blueprintName === undefined) {
+    return Blueprint.lookup(DEFAULT_BLUEPRINT_NAME, { paths });
+  }
+
+  let blueprint;
+
+  // Git URLs are unambiguous, so we can detect and install it now.
+  if (isGitRepo(blueprintName)) {
+    return downloadBlueprint(blueprintName, fromGit);
+  }
+
+  // npm package names, however, are *not* unambiguous because they could also
+  // be describing a path on disk. First, we will search the blueprint paths and
+  // prefer anything that we find locally.
+  try {
+    blueprint = Blueprint.lookup(blueprintName, { paths });
+  } catch (err) {
+    // If we didn't find it locally, we'll see if it's a valid npm package name
+    // and install it from npm if so.
+    if (maybeNpmPackage(blueprintName)) {
+      return downloadBlueprint(blueprintName, fromNpm);
     }
+
+    // Blueprint wasn't found and it couldn't be an npm package, so re-throw the
+    // original exception.
+
+    return Promise.reject(err);
   }
+
+  return Promise.resolve(blueprint);
+}
+
+function maybeNpmPackage(packageName) {
+  return validateNpmPackageName(packageName).validForNewPackages;
+}
+
+// Creates a temporary directory, runs a passed callback to install a blueprint
+// into that directory, then returns a blueprint loaded from the downloaded
+// directory.
+function downloadBlueprint(blueprintName, download) {
+  // Create a temporary directory
+  return mkdir('ember-cli')
+    // Download the blueprint into the temporary directory by calling the
+    // provided callback.
+    .then(tempPath => download(blueprintName, tempPath))
+    // The download callback returns a promise that resolves to the final path
+    // of the downloaded blueprint. Run `npm install` inside of it and then load
+    // it.
+    .then(blueprintPath =>
+      exec('npm install', { cwd: blueprintPath })
```

Comment:
I think we only need to do this for git repos, since `npm install <blueprintName>` will have resolved all the deps in the npm package case...